source/common/config/config_provider_impl.h

```diff
@@ -268,170 +296,69 @@ class ConfigSubscriptionInstance : public ConfigSubscriptionCommonBase {
-using ConfigSharedPtr = std::shared_ptr<Envoy::Config::ConfigProvider::Config>;
+protected:
+  /**
+   * Called when a new config proto is received via an xDS subscription.
+   * On successful validation of the config, must return a shared_ptr to a ConfigProvider::Config
+   * implementation that will be propagated to all mutable config providers sharing the
+   * subscription.
+   * Note that this function is called _once_ across all shared config providers per xDS
+   * subscription config update.
+   * @param config_proto supplies the configuration proto.
+   * @return ConfigConstSharedPtr the ConfigProvider::Config to share with other providers.
+   */
+  virtual ConfigProvider::ConfigConstSharedPtr
+  onConfigProtoUpdate(const Protobuf::Message& config_proto) PURE;
+};
 
 /**
  * Provides common subscription functionality required by ConfigProvider::ApiType::Delta DS APIs.
  */
 class DeltaConfigSubscriptionInstance : public ConfigSubscriptionCommonBase {
 protected:
-  DeltaConfigSubscriptionInstance(const std::string& name, const uint64_t manager_identifier,
-                                  ConfigProviderManagerImplBase& config_provider_manager,
-                                  TimeSource& time_source, const SystemTime& last_updated,
-                                  const LocalInfo::LocalInfo& local_info)
-      : ConfigSubscriptionCommonBase(name, manager_identifier, config_provider_manager, time_source,
-                                     last_updated, local_info) {}
-
+  using ConfigSubscriptionCommonBase::ConfigSubscriptionCommonBase;
   ~DeltaConfigSubscriptionInstance() override = default;
 
   /**
-   * Propagates a config update to the config providers and worker threads associated with the
-   * subscription.
-   *
-   * @param update_fn the callback to run on each worker thread.
+   * Must be called by the derived class' constructor.
+   * @param init_cb supplies an initial Envoy::Config::ConfigProvider::Config associated with the
+   * underlying subscription for each worker thread.
    */
-  void applyDeltaConfigUpdate(const std::function<void(const ConfigSharedPtr&)>& update_fn);
+  void initialize(const std::function<ConfigProvider::ConfigConstSharedPtr()>& init_cb) {
+    tls_->set([init_cb](Event::Dispatcher&) -> ThreadLocal::ThreadLocalObjectSharedPtr {
+      return std::make_shared<ThreadLocalConfig>(init_cb());
+    });
+  }
 };
 
 /**
  * Provides generic functionality required by the ConfigProvider::ApiType specific dynamic config
- * providers (see MutableConfigProviderBase and DeltaMutableConfigProviderBase).
+ * providers.
  *
  * This class can not be instantiated directly; instead, it provides the foundation for
  * dynamic config provider implementations which derive from it.
  */
 class MutableConfigProviderCommonBase : public ConfigProvider {
 public:
-  ~MutableConfigProviderCommonBase() override { subscription_->unbindConfigProvider(this); }
+  ~MutableConfigProviderCommonBase() override = default;
```

Comment:
do you need to specify the default here at all?

Reply:
I am not sure if there is an entry in style guide for this. I think it's a pattern in our code base that  subclasses have a defaulted destructor override:

grep "~.*override = default" -R source/ include/ test/ | wc
     73     366    6626

Reply:
I would skip it, I don't think it contributes anything.

Reply:
done